### ISSUE TITLE
How to Calculate Proper Scale/Shift Values When WF-VAE Output Range Differs from [-1,1] for Diffusion Model Input?

### ISSUE THREAD

**Issue by contain-mm** [PKU-YuanGroup/WF-VAE#17]
I sincerely apologize for reaching out again, but I find myself facing a particularly challenging issue that I can only resolve with your expert guidance. In reviewing your implementation, I noticed the scale and shift parameters adopt empirical values from Stable Diffusion (scale=0.18215). Having successfully replicated your WF-VAE architecture with my custom dataset - which I intend to use as a preprocessing component for a diffusion model - I've observed that the encoded data distribution doesn't conform to the expected [-1, 1] range.

This brings me to my fundamental question: Should I maintain the original empirical parameters, or would it be more appropriate to derive dataset-specific values using your vae_var_test.py utility? When executing this script on my data, I obtained mean=-0.005 and std=0.5739. Could you kindly advise whether these values fall within expected parameters? Most importantly, I would be deeply grateful for your guidance on the proper methodology to determine optimal scale and shift values in this context.

Please accept my sincere apologies for this repeated inquiry, and thank you most profoundly for considering my request. Your expertise would be invaluable in resolving this matter.

**Comment by qqingzheng** [comment]
😂😂 No need to apologize at all! I'm really happy to help you! I recommend running vae_var_test.py on a large-scale dataset to obtain the mean and std of latents. Also, it's best to calculate the mean and std separately for each latent channel to ensure proper normalization of the latents.

You can use the measured mean and std to verify whether the latents are properly normalized—this may serve as a validation method.

**Comment by contain-mm** [comment]
好的，作者您人真是太好了！！！非常感谢您详细的建议，但是我还有一点不是特别清楚。就是这个大规模的数据集指的是训练+测试+验证吗？还是只需要训练用的数据集就行了？下面是我用训练时使用的训练数据集（滑动步长为2）运行vae_var_test.py的代码以及结果。我还有几个问题咨询您：

1、我看您的代码中判别器的开始步数是从5开始的，这个参数是您真实使用的吗？因为我现在复现WF-VAE，让它重建5帧(数据形状为（b,t,c,h,w）),让生成器损失稳定后，大概30w步才开始加入判别器，总共训练了100w步，后面重建的图像还可以，不知道我这样操作有没有问题，因为训练的步数太长了（我使用的是单卡4090，内存为50G）。
2、我这个数据的通道是单通道，所以我直接关闭了感知损失，让感知损失的前面的系数为0，这是否会对模型造成性能上的影响？

感谢您百忙之中的回复，万分感谢！

@torch.no_grad()
def vae_var(configs):

    data_type = torch.bfloat16
    device='cuda:0'
    # ---- Load Model ----
    vae = WFVAE2Model()
    vae.init_from_ckpt("************")
    vae = vae.to(device).to(data_type)
    vae.eval()

    # ---- Prepare Dataset ----
    train_dataloader = DataLoader(ShangHaiDataset(configs["train_data_path"], True, configs),
                                       batch_size=configs["train_batch_size"],
                                       shuffle=True,
                                       drop_last=True,
                                       num_workers=4)
    # valid_dataloader = DataLoader(ShangHaiDataset(configs["valid_data_path"], False, configs),
    #                                    batch_size=configs["test_batch_size"],
    #                                    shuffle=False,
    #                                    drop_last=True,
    #                                    num_workers=4)
    # ---- Inference ----
    latents_list = []
    mean_list = []
    std_list = []

    for batch_id,batch in enumerate(tqdm(train_dataloader)):
        x = rearrange(batch,'b t c h w -> b c t h w')
        shift = torch.tensor(configs["shift"])[None, :, None, None, None].to(device=device, dtype=data_type)
        scale = torch.tensor(configs["scale"])[None, :, None, None, None].to(device=device, dtype=data_type)
        x = x.to(device=device, dtype=data_type)  # b c t h w
        latents = vae.encode(x)[0].sample()
        # latents=(latents-shift)*scale
        # print(
        #     f"latents range: [{latents.min().item():.4f}, {latents.max().item():.4f}]\n"
        #     f"mean: {latents.mean().item():.4f}, std: {latents.std().item():.4f}"
        # )
        # recon_latents = latents / scale + shift

        # test_recon = vae.decode(recon_latents).sample
        # print(
        #     f"test_recon range: [{test_recon.min().item():.4f}, {test_recon.max().item():.4f}]\n"
        #     f"mean: {test_recon.mean().item():.4f}, std: {test_recon.std().item():.4f}"
        # )
        # test_recon = test_recon.clamp(0, 1)
        means = torch.mean(latents)
        stds = torch.std(latents)
        latents_list.append(latents.cpu())
        mean_list.append(means.item())
        std_list.append(stds.item())

    print("mean:", torch.mean(torch.tensor(mean_list)))
    print("std:", torch.mean(torch.tensor(std_list)))
    # 最终统计输出
    print(f"各通道极值范围:")
    all_latents = torch.cat(latents_list)
    means = torch.mean(all_latents, dim=(0, 2, 3, 4))
    stds = torch.std(all_latents, dim=(0, 2, 3, 4))
    print("mean:", means)
    print("std:", stds)
    print(f'normalizer={1 / stds}')

    # normalizer = 1 / std

if __name__ == "__main__":
    config=Tools.parse_yaml("./config.yaml")
    vae_var(config)

There are 15225 sequences.
100%|███████████████████████████████████████| 2537/2537 [11:19<00:00,  3.73it/s]
mean: tensor(-0.0050)
std: tensor(0.5739)
各通道极值范围:
mean: tensor([ 0.0281,  0.0454, -0.0115, -0.1475,  0.0630,  0.1074,  0.1953, -0.0928,
         0.0913,  0.0374, -0.0128, -0.1108,  0.0055,  0.1016, -0.1099, -0.2256],
       dtype=torch.bfloat16)
std: tensor([0.6641, 0.4219, 0.4199, 0.4648, 0.3496, 0.9922, 0.5977, 1.0938, 0.4902,
        0.4258, 0.3027, 0.3750, 0.3047, 0.3496, 0.5625, 0.5664],
       dtype=torch.bfloat16)
normalizer=tensor([1.5078, 2.3750, 2.3750, 2.1562, 2.8594, 1.0078, 1.6719, 0.9141, 2.0469,
        2.3438, 3.2969, 2.6719, 3.2812, 2.8594, 1.7812, 1.7656],
       dtype=torch.bfloat16)

Process finished with exit code 0

**Comment by contain-mm** [comment]
这是我用上述运行得到的各通道的mean和std对latents进行归一化得到的结果，感觉很奇怪，它并没有在[-1,1]这个范围。

    for batch_id,batch in enumerate(tqdm(train_dataloader)):
        x = rearrange(batch,'b t c h w -> b c t h w')
        shift = torch.tensor(configs["shift"])[None, :, None, None, None].to(device=device, dtype=data_type)
        scale = torch.tensor(configs["scale"])[None, :, None, None, None].to(device=device, dtype=data_type)
        x = x.to(device=device, dtype=data_type)  # b c t h w
        latents = vae.encode(x)[0].sample()
        print(
            f"latents range: [{latents.min().item():.4f}, {latents.max().item():.4f}]\n"
            f"mean: {latents.mean().item():.4f}, std: {latents.std().item():.4f}"
        )
        latents=(latents-shift)*scale
        print(
            f"latents range: [{latents.min().item():.4f}, {latents.max().item():.4f}]\n"
            f"mean: {latents.mean().item():.4f}, std: {latents.std().item():.4f}"
        )
        recon_latents = latents / scale + shift

        test_recon = vae.decode(recon_latents).sample
        print(
            f"test_recon range: [{test_recon.min().item():.4f}, {test_recon.max().item():.4f}]\n"
            f"mean: {test_recon.mean().item():.4f}, std: {test_recon.std().item():.4f}"
        )

There are 15225 sequences.
  0%|                                                  | 0/2537 [00:00<?, ?it/s]latents range: [-6.7812, 6.4062]
mean: 0.0002, std: 0.6094
latents range: [-7.3125, 6.2500]
mean: 0.0011, std: 0.4785
test_recon range: [-0.0659, 0.7070]
mean: 0.0337, std: 0.1011
  0%|                                        | 1/2537 [00:02<1:41:09,  2.39s/it]latents range: [-7.0938, 6.6562]
mean: -0.0200, std: 0.4434
latents range: [-7.6562, 6.5000]
mean: -0.0084, std: 0.3359
test_recon range: [-0.0654, 0.7305]
mean: 0.0162, std: 0.0713

**Comment by qqingzheng** [comment]
1. 你在30w步启动GAN是完全没问题的。
2. 舍弃了lpips损失我觉得影响不大，GAN其实和lpips的作用会有点类似。
3. 关于这个latents归一化的问题，我看着代码感觉没问题，你可以加我微信chestnutlzj细聊。

**Comment by contain-mm** [comment]
好的，谢谢您清楚的答复！非常荣幸，已加🤭
